platform/darwin/src/MGLOfflineStorage.h

```diff
@@ -238,6 +238,31 @@ typedef void (^MGLOfflinePackRemovalCompletionHandler)(NSError * _Nullable error
  */
 @property (nonatomic, readonly) unsigned long long countOfBytesCompleted;
 
+/**
+ Sets a URL template for the given scheme.
+
+ Any URLs encountered in styles and subsequently loaded sources will be fed through
+ a URL replacement mechanism that e.g. transforms `mapbox://` URLs into regular
+ `https://` URLs. You can provide URL replacement patterns for your own scheme,
+ e.g. `example://`.
+
+ Any query string will be preserved from the original URL and merged with a
+ query string of the template URL, if any.
+
+ @param tpl A templated URL. it may contain these brace-enclosed replacement
+    tokens, e.g. for `"http://example.com/foo/bar/baz.png"`, valid tokens are:
+    * `{scheme}`: `"http"`
+    * `{domain}`: `"example.com"`
+    * `{path}`: `"foo/bar/baz.png"` (without leading `/`)
+    * `{directory}`: `"foo/bar/"` (with trailing `/`)
+    * `{filename}`: `"baz"` (without file extension)
+    * `{extension}`: `".png"` (optionally with a preceding `@2x` ratio specifier)
```

Comment:
Mini-languages in strings are un-Cocoa. Instead of exposing yet another mini-language that resembles but works differently than `textField` tokens in the runtime styling API, have you considered a lambda-based (block-based) API? The developer could specify a block (or perhaps an NSValueTransformer) that accepts an NSURL and returns a transformed NSURL. Within the block, they could easily parse and manipulate the URL via NSURLComponents.

Even more natural would be a delegate pattern: the developer could have their application delegate conform to a new MGLOfflineStorageDelegate protocol by implementing an `-offlineStorage:transformedURLForTileURL:` method.